File: csharp/queen-attack/QueenAttack.cs
using System;

public static class QueenAttack
{
    public static bool CanAttack((int Row, int Column) white, (int Row, int Column) black) =>
        Math.Abs(white.Row - black.Row) == Math.Abs(white.Column - black.Column)
        || white.Row == black.Row
        || white.Column == black.Column;

    public static (int Row, int Column) Create(int row, int column) =>
        (row, column) switch
        {
            var (r, _) when r < 0 || r > 7 =>
                throw new ArgumentOutOfRangeException(nameof(row), "must be between 0 and 7"),
            var (_, c) when c < 0 || c > 7 =>
                throw new ArgumentOutOfRangeException(nameof(column), "must be between 0 and 7"),
            _ => (row, column)
        };
}

File: fsharp/queen-attack/QueenAttack.fs
module QueenAttack

let create (col: int, row: int) = 
    // A) 0 <= col && col < 8 && 0 <= row && row < 8
    // B) Seq.contains col [0..7] && Seq.contains row [0..7] 
    [col; row]
    |> Seq.forall (fun x -> Seq.contains x [0..7])

let canAttack (queen1: int * int) (queen2: int * int) = 
    
    if fst queen1 = fst queen2 then
        true
    elif snd queen1 = snd queen2 then
        true
    elif abs (fst queen1 - fst queen2) = abs (snd queen1 - snd queen2) then 
        true
    else
        false